tests/unit/test_refine.py

```diff
@@ -38,237 +42,56 @@
 
-    def fit_to_size(self, coreset_size: int) -> None:
+    def fit_to_size(self, coreset_size: int):
         raise NotImplementedError
 
 
-class TestRefine(unittest.TestCase):
-    """
-    Tests related to refine.py functions.
-    """
+class TestRefine:
+    """Tests related to `coreax.refine.Refine`."""
 
-    def setUp(self):
-        self.random_key = random.key(0)
+    random_key = jr.key(0)
 
-    def test_validate_coreset_ok(self) -> None:
+    # Disable pylint warning for protected-access as we are testing a single part of
+    # the over-arching algorithm ``coreax.refine.Refine._validate_coreset``.
+    # pylint: disable=protected-access
+    def test_validate_coreset_ok(self):
         """Check validation passes with populated coresubset."""
         coreset = CoresetMock()
         coreset.original_data = coreax.data.ArrayData.load(1)
         coreset.coreset = jnp.array(1)
         coreset.coreset_indices = jnp.array(0)
-        # Disable pylint warning for protected-access as we are testing a single part of
-        # the over-arching algorithm
-        # pylint: disable=protected-access
         coreax.refine.Refine._validate_coreset(coreset)
-        # pylint: enable=protected-access
 
-    def test_validate_coreset_no_fit(self) -> None:
+    def test_validate_coreset_no_fit(self):
         """Check validation fails when coreset has not been calculated."""
         coreset = CoresetMock()
         coreset.original_data = coreax.data.ArrayData.load(1)
-        # Disable pylint warning for protected-access as we are testing a single part of
-        # the over-arching algorithm
-        # pylint: disable=protected-access
-        self.assertRaises(
-            coreax.util.NotCalculatedError,
-            coreax.refine.Refine._validate_coreset,
-            coreset,
-        )
-        # pylint: enable=protected-access
-
-    def test_validate_coreset_not_coresubset(self) -> None:
+        with pytest.raises(coreax.util.NotCalculatedError):
+            coreax.refine.Refine._validate_coreset(coreset)
+
+    def test_validate_coreset_not_coresubset(self):
         """Check validation raises TypeError when not a coresubset."""
         coreset = CoresetMock()
         coreset.original_data = coreax.data.ArrayData.load(1)
         coreset.coreset = jnp.array(1)
-        # Disable pylint warning for protected-access as we are testing a single part of
-        # the over-arching algorithm
-        # pylint: disable=protected-access
-        self.assertRaises(TypeError, coreax.refine.Refine._validate_coreset, coreset)
-        # pylint: enable=protected-access
-
-
-class TestRefineRegular(unittest.TestCase):
-    """
-    Tests related to :meth:`~coreax.refine.RefineRegular`.
-    """
-
-    def setUp(self):
-        self.random_key = random.key(0)
-
-    def test_refine_ones(self) -> None:
-        """
-        Test that refining an optimal coreset leaves ``coreset_indices`` unchanged.
```

Comment:
I don't think we're actually testing this any more?

Reply:
So I wasn't sure what this test was trying to check. As far as I'm aware, we don't expect refine to have the machinery to shortcut the expensive parts of the refinement step if `coreset_indices = best_indices`. I.E. we still expect refinement to proceed as usual even if we started with the "correct" coreset indices. In this test we have multiple feasible values for the `best_indices`, `{0, 1}` or `{0, 3}` or `{1, 2}` or `{2,3}`, so are we expecting refine to enumerate all of the possible `best_indices` and only return the original coreset indices if they were one of the possible `best_indices`?

I.E. I don't think this test is really checking anything different to `test_refine_ints`.

Reply:
The difference is that in this test we start with an already-optimal coreset, and we're checking that it's not changed (i.e. to a different optimal coreset). If we guarantee this stability/idempotency property, then it makes sense to have this test in place; if we don't, the current implementation as a parametrisation of test_refine (which doesn't check that the coreset_indices are unchanged) is OK

Reply:
I can not identify any functionality within the existing refine methods that would guarantee the idempotency property, nor can I find any documentation suggesting that this is a required/expected behaviour (excluding the old test), so I believe the current parameterisation of `test_refine` should suffice.